rust/src/main.rs

```diff
@@ -0,0 +1,339 @@
+#![feature(test)]
+#![deny(unsafe_code)]
+
+extern crate itertools;
+extern crate mersenne_twister;
+extern crate num;
+extern crate rand;
+extern crate test;
+
+// Use BLAS directly
+#[cfg(feature = "direct_blas")]
+extern crate cblas;
+
+#[cfg(feature = "direct_blas")]
+extern crate blas_src;
+
+// Use ndarray (with BLAS implementation)
+#[cfg(not(feature = "direct_blas"))]
+#[macro_use(s)]
+extern crate ndarray;
+
+use std::time::{Duration, Instant};
+use std::u32;
+use std::fs::OpenOptions;
+use std::io::{BufWriter, Write};
+
+use test::black_box;
+use num::complex::Complex64;
+use rand::Rng;
+
+mod util;
+use util::{gen_rng, myrand};
+
+#[cfg(feature = "direct_blas")]
+mod direct_blas;
+#[cfg(feature = "direct_blas")]
+use direct_blas::{randmatstat, randmatmul, check_randmatmul};
+
+#[cfg(not(feature = "direct_blas"))]
+use ndarray::Array2;
+#[cfg(not(feature = "direct_blas"))]
+use util::fill_rand;
+#[cfg(not(feature = "direct_blas"))]
+use num::Zero;
+
+const NITER: u32 = 5;
+
+#[cfg(not(feature = "direct_blas"))]
+fn nrand<R: Rng>(shape: (usize, usize), rng: &mut R) -> Array2<f64> {
+    let mut m = Array2::zeros(shape);
+    fill_rand(&mut m, rng);
+    m
+}
+
+fn fib(n: i32) -> i32 {
+    let n = black_box(n); // prevent over-optimization
+    if n < 2 {
+        n
+    } else {
+        fib(n - 1) + fib(n - 2)
+    }
+}
+
+fn mandel(z: Complex64) -> u32 {
+    use std::iter;
+
+    iter::repeat(z)
+        .scan(z, |z, c| {
+            let current = *z;
+            *z = current * current + c;
+            Some(current)
+        })
+        .take(80)
+        .take_while(|z| z.norm() <= 2.0)
+        .count() as u32
```

Comment:
Is this kind of iterator-composition code really the fastest way to do it in Rust?  Again, wouldn't you just write a loop?

Reply:
I can't speak to this code in particular, or whether the style is ideal for performance, but I have noticed that this general style is very common in Rust.

Reply:
As already mentioned by @Boiethios, this is pretty idiomatic and performant. Previously, I had a version of this function without an extensive use of the Iterator API [(here)](https://github.com/Enet4/julia-bench-rs/blame/5d71e1a45437ba0c55280f7efe0b774125052a07/src/main.rs#L50-L61), and the subsequent changes to what we have now actually improved performance a tiny bit.